gh-85453: Adapt datetime.rst to devguide recommendations for code snippets and variables (#118068)
python/cpython: Doc/library/datetime.rst

   :noindex:

   An idealized time, independent of any particular day, assuming that every day

>    has exactly ``24*60*60`` seconds.  (There is no notion of "leap seconds" here.)

Reviewer: Let's revert this before landing:

```suggestion
   has exactly 24\*60\*60 seconds.  (There is no notion of "leap seconds" here.)
```

Reviewer: @uatach, I'll take the liberty to update this; hope you don't mind!

>    has exactly ``24 * 60 * 60`` seconds.  (There is no notion of "leap seconds" here.)

Reviewer 2: Why does this need code formatting? The existing output looks fine (https://docs.python.org/3.13/library/datetime.html#available-types).

Author: I just ported from old PR but I think it's better, maybe could remove the spaces.

   Only *days*, *seconds* and *microseconds* are stored internally.
   Arguments are converted to those units:

>    * A *millisecond* is converted to ``1000`` *microseconds*.

Reviewer 2: None of these changes look necessary.

Author: I think the italics call attention to the change in the time unit which is the subject of the previous paragraph.

>    * A *millisecond* is converted to 1000 *microseconds*.
>    * A *minute* is converted to 60 *seconds*.
>    * An *hour* is converted to 3600 *seconds*.
>    * A *week* is converted to 7 *days*.

Reviewer: This change does not seem right to me; this list describes unit conversions only.

```suggestion
   * A millisecond is converted to 1000 microseconds.
   * A minute is converted to 60 seconds.
   * An hour is converted to 3600 seconds.
   * A week is converted to 7 days.
```

   The smallest possible difference between non-equal :class:`timedelta` objects,
   ``timedelta(microseconds=1)``.

> Note that, because of normalization, ``timedelta.max > -timedelta.min`` is true.

Reviewer 2: This might be clearer as:
```suggestion
Note that, because of normalization, ``timedelta.max`` is greater than ``-timedelta.min``.
```
But I think the existing phrasing is fine too.

+--------------------------------+-----------------------------------------------+
| Operation                      | Result                                        |
+================================+===============================================+

> | ``t1 = t2 + t3``               | Sum of *t2* and *t3*. Afterwards ``t1 - t2 == |

Reviewer: Double backquotes should be used for variable names, as per [the devguide style recommendations](https://devguide.python.org/documentation/markup/#inline-markup).

```suggestion
| ``t1 = t2 + t3``               | Sum of ``t2`` and ``t3``. Afterwards ``t1 - t2 == |
```

|                                | t3`` and ``t1 - t3 == t2`` are true. (1)      |
+--------------------------------+-----------------------------------------------+

> | ``t1 = t2 - t3``               | Difference of *t2*  and *t3*. Afterwards ``t1 |

Reviewer: Ditto:

```suggestion
| ``t1 = t2 - t3``               | Difference of ``t2` and ``t3``. Afterwards ``t1 |
```

+--------------------------------+-----------------------------------------------+
| Operation                      | Result                                        |
+================================+===============================================+

> | ``t1 = t2 + t3``               | Sum of ``t2`` and ``t3``. Afterwards ``t1 - t2|
> |                                |  == t3`` and ``t1 - t3 == t2`` are true. (1)  |

Reviewer: Sphinx lint is complaining. You need to format it similar to this:

```suggestion
| ``t1 = t2 + t3``               | Sum of ``t2`` and ``t3``.                     |
|                                | Afterwards ``t1 - t2 == t3`` and              |
|                                | ``t1 - t3 == t2`` are true. (1)               |
```

+--------------------------------+-----------------------------------------------+
| ``t1 = t2 - t3``               | Difference of ``t2``  and ``t3``. Afterwards  |
|                                | ``t1 == t2 - t3`` and ``t2 == t1 + t3`` are   |
|                                | true. (1)(6)                                  |
+--------------------------------+-----------------------------------------------+
| ``t1 = t2 * i or t1 = i * t2`` | Delta multiplied by an integer.               |
|                                | Afterwards ``t1 // i == t2`` is true,         |
|                                | provided ``i != 0``.                          |

> +--------------------------------+-----------------------------------------------+

Reviewer 3: Is this line needed?

Reviewer: `t1` and `t2` should be marked up as variables, not as parameters; according to the style recommendations of the devguide, variables, code snippets, and literals are marked up using double backquotes.

Reviewer 3: It is not about marking `t1` and `t2`. I asked about the meaning of this line. What does it separate?

Although it may be a separate issue.

Reviewer: It is a separate issue and can be dealt with in a separate PR.

|                                | In general, ``t1  * i == t1 * (i-1) + t1``    |
|                                | is true. (1)                                  |
+--------------------------------+-----------------------------------------------+
| ``t1 = t2 * f or t1 = f * t2`` | Delta multiplied by a float. The result is    |
|                                | rounded to the nearest multiple of            |
|                                | timedelta.resolution using round-half-to-even.|
+--------------------------------+-----------------------------------------------+
| ``f = t2 / t3``                | Division (3) of overall duration ``t2`` by    |
|                                | interval unit ``t3``. Returns a :class:`float`|
|                                | object.                                       |
+--------------------------------+-----------------------------------------------+

> | ``t1 = t2 / f or t1 = t2 / i`` | Delta divided by a float or an int. The result|

Reviewer 3: Needed a reference to `(3)`.

Reviewer: Added to the issue: https://github.com/python/cpython/issues/85453#issuecomment-2075721997

| ``+t1``                        | Returns a :class:`timedelta` object with the  |
|                                | same value. (2)                               |
+--------------------------------+-----------------------------------------------+
| ``-t1``                        | equivalent to ``timedelta(-t1.days,           |
|                                | -t1.seconds*, -t1.microseconds)``,            |
|                                | and to ``t1 * -1``. (1)(4)                    |
+--------------------------------+-----------------------------------------------+

> | ``abs(t)``                     | equivalent to +\ *t* when ``t.days >= 0``,    |
> |                                | and to -*t* when ``t.days < 0``. (2)          |

Reviewer: ```suggestion
| ``abs(t)``                     | equivalent to ``t`` when ``t.days >= 0``,    |
|                                | and to ``-t`` when ``t.days < 0``. (2)          |
```

| ``+t1``                        | Returns a :class:`timedelta` object with the  |
|                                | same value. (2)                               |
+--------------------------------+-----------------------------------------------+
| ``-t1``                        | equivalent to ``timedelta(-t1.days,           |
|                                | -t1.seconds*, -t1.microseconds)``,            |
|                                | and to ``t1 * -1``. (1)(4)                    |
+--------------------------------+-----------------------------------------------+

> | ``abs(t)``                     | equivalent to ``t`` when ``t.days >= 0``,     |

Reviewer 3: ```suggestion
| ``abs(t)``                     | equivalent to ``+t`` when ``t.days >= 0``,    |
```
BTW, there is a bug in the Python implementation.

Reviewer: Remark was addressed; marking this as resolved.

| ``+t1``                        | Returns a :class:`timedelta` object with the  |
|                                | same value. (2)                               |
+--------------------------------+-----------------------------------------------+

> | ``-t1``                        | equivalent to ``timedelta(-t1.days,           |

Reviewer 3: Should not the sentence start with the title case?

Reviewer: It is a separate issue and should be dealt with in a separate PR. This PR does not need to grow more.

Reviewer: Added: https://github.com/python/cpython/issues/85453#issuecomment-2075718920

   This is exact and cannot overflow.

(3)

>    Division by zero raises :exc:`ZeroDivisionError`.

Reviewer 3: I understand what it means, but it would be nice to explain what does "zero" mean. Is it already explained anywhere?

Reviewer: Can you add this remark to the issue so we can follow it up?

Reviewer: Added: https://github.com/python/cpython/issues/85453#issuecomment-2075715789

   This is exact and cannot overflow.

(3)

>    Division by ``0`` raises :exc:`ZeroDivisionError`.

Reviewer 2: This doesn't need code formatting.

Author: I'm not sure why some numbers don't need formatting

Reviewer 2: I don't think any of them need formatting.

Reviewer: The devguide states that [literals are marked up using double backticks](https://devguide.python.org/documentation/markup/#inline-markup):

> backquotes: \`\`text\`\` for code samples, variables, and literals.

Reviewer: But for this particular example, I'd definitely spell it out:

```suggestion
   Division by zero raises :exc:`ZeroDivisionError`.
```

.. classmethod:: date.fromordinal(ordinal)

   Return the date corresponding to the proleptic Gregorian ordinal, where

>    January 1 of year 1 has ordinal ``1``.

Reviewer: I'm not sure about this one; IMO it's best kept unchanged.

```suggestion
   January 1 of year 1 has ordinal 1.
```

Reviewer 2: I don't think simple integers should receive code markup. It would look distracting.

Reviewer: Only literals should receive markup; I added this remark since I don't think it qualifies as a code literal.

.. attribute:: date.month

>    Between ``1`` and ``12`` inclusive.

Reviewer 3: This is not a code. They are just numbers.

Reviewer: The change has been reverted; marking as resolved.

.. attribute:: date.day

>    Between ``1`` and the number of days in the given month of the given year.

Reviewer 3: Not a code.

Reviewer: The change has been reverted; marking as resolved.

   ``timedelta.seconds`` and ``timedelta.microseconds`` are ignored.

(3)
   This is exact, and cannot overflow. ``timedelta.seconds`` and

>    ``timedelta.microseconds`` are ``0``, and ``date2 + timedelta == date1`` after.

Reviewer 2: Remove code formatting for ``0``

   ``timedelta.seconds`` and ``timedelta.microseconds`` are ignored.

(3)

>    This is exact, and cannot overflow. ``timedelta.seconds`` and

Reviewer 2: This is a good change, but 0 on the next line doesn't need code formatting.

Reviewer: I wonder if some of these `timedelta.<attr>` markups should use `:attr:` roles instead of the double backquotes used for variable names, code examples, and literals.

.. method:: date.toordinal()

   Return the proleptic Gregorian ordinal of the date, where January 1 of year 1

>    has ordinal ``1``. For any :class:`date` object *d*,

Reviewer: This is not a code literal.

```suggestion
   has ordinal 1. For any :class:`date` object *d*,
```

   ``date.fromordinal(d.toordinal()) == d``.

.. method:: date.weekday()

>    Return the day of the week as an integer, where Monday is ``0`` and Sunday is ``6``.

Reviewer: Ditto

```suggestion
   Return the day of the week as an integer, where Monday is 0 and Sunday is 6.
```

   For example, ``date(2002, 12, 4).weekday() == 2``, a Wednesday. See also
   :meth:`isoweekday`.

.. method:: date.isoweekday()

>    Return the day of the week as an integer, where Monday is ``1`` and Sunday is ``7``.

Reviewer: Ditto

```suggestion
   Return the day of the week as an integer, where Monday is 1 and Sunday is 7.
```

.. method:: date.strftime(format)

   Return a string representing the date, controlled by an explicit format string.

>    Format codes referring to hours, minutes or seconds will see ``0`` values.

Reviewer 2: Same here

Like a :class:`date` object, :class:`.datetime` assumes the current Gregorian
calendar extended in both directions; like a :class:`.time` object,

> :class:`.datetime` assumes there are exactly ``3600 * 24`` seconds in every day.

Reviewer: Restore the previous whitespace formatting.

```suggestion
:class:`.datetime` assumes there are exactly ``3600*24`` seconds in every day.
```

Like a :class:`date` object, :class:`.datetime` assumes the current Gregorian
calendar extended in both directions; like a :class:`.time` object,

> :class:`.datetime` assumes there are exactly ``3600*24`` seconds in every day.

Reviewer 2: This change isn't needed.

      failure.

   .. versionchanged:: 3.6

>       :meth:`fromtimestamp` may return instances with :attr:`.fold` set to ``1``.

Reviewer 2: Not needed

.. classmethod:: datetime.fromordinal(ordinal)

   Return the :class:`.datetime` corresponding to the proleptic Gregorian ordinal,

>    where January 1 of year 1 has ordinal ``1``. :exc:`ValueError` is raised unless

Reviewer: I'm not sure about this one; perhaps it is best left unchanged.

```suggestion
   where January 1 of year 1 has ordinal 1. :exc:`ValueError` is raised unless
```

.. classmethod:: datetime.fromordinal(ordinal)

   Return the :class:`.datetime` corresponding to the proleptic Gregorian ordinal,
   where January 1 of year 1 has ordinal 1. :exc:`ValueError` is raised unless
   ``1 <= ordinal <= datetime.max.toordinal()``. The hour, minute, second and

>    microsecond of the result are all ``0``, and :attr:`.tzinfo` is ``None``.

Reviewer 2: Not needed

.. attribute:: datetime.month

>    Between ``1`` and ``12`` inclusive.

Reviewer: These are not code literals.

```suggestion
   Between 1 and 12 inclusive.
```

.. attribute:: datetime.day

>    Between ``1`` and the number of days in the given month of the given year.

Reviewer: ```suggestion
   Between 1 and the number of days in the given month of the given year.
```

   In ``[0, 1]``. Used to disambiguate wall times during a repeated interval. (A
   repeated interval occurs when clocks are rolled back at the end of daylight saving
   time or when the UTC offset for the current zone is decreased for political reasons.)

>    The value ``0`` (1) represents the earlier (later) of the two moments with the same wall

Reviewer 3: Why is `0` formatted, but `1` is not?

Reviewer: Is it a reference to L1195? There is a similar reference later in the docs. `0` is marked up since it is a literal.

Author: Reading again, I think the `1` is a number and not a reference... `0` is earlier and `1` is later

Reviewer: Correct. See also PEP-495.

>    The value ``0`` (``1``) represents the earlier (later) of the two moments with the same wall

Reviewer: ```suggestion
   The values ``0`` and ``1`` respectively represents the earlier and later of the two moments with the same wall
```

   In ``[0, 1]``. Used to disambiguate wall times during a repeated interval. (A
   repeated interval occurs when clocks are rolled back at the end of daylight saving
   time or when the UTC offset for the current zone is decreased for political reasons.)

>    The values ``0`` and ``1`` represent, respectively, the earlier and later of the two

Reviewer 2: ```suggestion
   The values 0 and 1 represent, respectively, the earlier and later of the two
```

+---------------------------------------+--------------------------------+

(1)

>    ``datetime2`` is a duration of timedelta removed from ``datetime1``, moving forward in

Reviewer 3: ```suggestion
   ``datetime2`` is a duration of ``timedelta`` removed from ``datetime1``, moving forward in
```
It is a variable name here.

(1)
   datetime2 is a duration of timedelta removed from datetime1, moving forward in
   time if ``timedelta.days > 0``, or backward if ``timedelta.days < 0``. The
   result has the same :attr:`~.datetime.tzinfo` attribute as the input datetime, and
   ``datetime2 - datetime1 == timedelta`` after. :exc:`OverflowError` is raised if
   ``datetime2.year`` would be smaller than :const:`MINYEAR` or larger than
   :const:`MAXYEAR`. Note that no time zone adjustments are done even if the
   input is an aware object.

(2)

>    Computes the datetime2 such that ``datetime2 + timedelta == datetime1``. As for

Reviewer 3: `datetime2` should be formatted.

   object *t* such that ``datetime2 + t == datetime1``. No time zone adjustments
   are done in this case.

>    If both are aware and have different :attr:`~.datetime.tzinfo` attributes, ``a - b`` acts

Reviewer: Please revert unrelated style changes like this.

```suggestion
   If both are aware and have different :attr:`~.datetime.tzinfo` attributes, ``a-b`` acts
```

.. method:: datetime.utctimetuple()

   If :class:`.datetime` instance *d* is naive, this is the same as
   ``d.timetuple()`` except that :attr:`~.time.struct_time.tm_isdst` is forced to ``0`` regardless of what
   ``d.dst()`` returns. DST is never in effect for a UTC time.

   If *d* is aware, *d* is normalized to UTC time, by subtracting
   ``d.utcoffset()``, and a :class:`time.struct_time` for the
   normalized time is returned. :attr:`!tm_isdst` is forced to ``0``. Note

>    that an :exc:`OverflowError` may be raised if ``d.year`` was

Reviewer 3: Would not be better to use ``` :samp:`{d}.year` ```? And in other similar cases.

Reviewer: I think double backquotes is better in this case, but I don't feel strongly about it.

.. method:: datetime.utctimetuple()

   If :class:`.datetime` instance *d* is naive, this is the same as

>    ``d.timetuple()`` except that :attr:`~.time.struct_time.tm_isdst` is forced to ``0`` regardless of what

Reviewer 2: Not needed

   ``d.dst()`` returns. DST is never in effect for a UTC time.

   If *d* is aware, *d* is normalized to UTC time, by subtracting
   ``d.utcoffset()``, and a :class:`time.struct_time` for the

>    normalized time is returned. :attr:`!tm_isdst` is forced to ``0``. Note

Reviewer 2: ```suggestion
   normalized time is returned. :attr:`!tm_isdst` is forced to 0. Note
```

.. method:: datetime.weekday()

>    Return the day of the week as an integer, where Monday is ``0`` and Sunday is ``6``.

Reviewer: ```suggestion
   Return the day of the week as an integer, where Monday is 0 and Sunday is 6.
```

   The same as ``self.date().weekday()``. See also :meth:`isoweekday`.

.. method:: datetime.isoweekday()

>    Return the day of the week as an integer, where Monday is ``1`` and Sunday is ``7``.

Reviewer: ```suggestion
   Return the day of the week as an integer, where Monday is 1 and Sunday is 7.
```

   Return a string representing the date and time in ISO 8601 format:

   - ``YYYY-MM-DDTHH:MM:SS.ffffff``, if :attr:`microsecond` is not ``0``
   - ``YYYY-MM-DDTHH:MM:SS``, if :attr:`microsecond` is ``0``

   If :meth:`utcoffset` does not return ``None``, a string is
   appended, giving the UTC offset:

   - ``YYYY-MM-DDTHH:MM:SS.ffffff+HH:MM[:SS[.ffffff]]``, if :attr:`microsecond`

>      is not 0

Reviewer: ```suggestion
     is not ``0``
```

   Return a string representing the date and time in ISO 8601 format:

>    - ``YYYY-MM-DDTHH:MM:SS.ffffff``, if :attr:`microsecond` is not ``0``

Reviewer 2: Avoid code formatting for any "0" or "1"

   * ``fold in [0, 1]``.

   If an argument outside those ranges is given, :exc:`ValueError` is raised. All

>    default to ``0`` except *tzinfo*, which defaults to :const:`None`.

Reviewer 2: Why was this link removed?

Reviewer: This remark is outdated; marking as resolved.

   In ``[0, 1]``. Used to disambiguate wall times during a repeated interval. (A
   repeated interval occurs when clocks are rolled back at the end of daylight saving
   time or when the UTC offset for the current zone is decreased for political reasons.)

>    The value ``0`` (1) represents the earlier (later) of the two moments with the same wall

Reviewer: ```suggestion
   The values ``0`` and ``1`` respectively represents the earlier and later of the two moments with the same wall
```

   In ``[0, 1]``. Used to disambiguate wall times during a repeated interval. (A
   repeated interval occurs when clocks are rolled back at the end of daylight saving
   time or when the UTC offset for the current zone is decreased for political reasons.)

>    The value ``0`` (1) represents the earlier (later) of the two moments with the same wall
>    time representation.

Reviewer: ```suggestion
   The values ``0`` and ``1`` represent, respectively, the earlier and later of the two
   moments with the same wall time representation.
```

|           |                                |  Samstag (de_DE)       |       |
+-----------+--------------------------------+------------------------+-------+
| ``%w``    | Weekday as a decimal number,   | 0, 1, ..., 6           |       |

> |           | where ``'0'`` is Sunday        |                        |       |
> |           | and ``'6'`` is Saturday.       |                        |       |

Reviewer: I'm not sure about this one; consider leaving it unchanged.

```suggestion
|           | where 0 is Sunday        |                        |       |
|           | and 6 is Saturday.       |                        |       |
```

> |           | and 6 is Saturday.             |                        |       |

Reviewer 3: What is changed?

Reviewer: This remark is outdated; marking it as resolved.

|           | decimal number. All days in a  |                        |       |
|           | new year preceding the first   |                        |       |
|           | Sunday are considered to be in |                        |       |

> |           | week ``'0'``.                  |                        |       |

Reviewer: Not a literal.

```suggestion
|           | week 0.                  |                        |       |
```

+-----------+--------------------------------+------------------------+-------+
| ``%W``    | Week number of the year        | 00, 01, ..., 53        | \(7), |
|           | (Monday as the first day of    |                        | \(9)  |
|           | the week) as a zero-padded     |                        |       |
|           | decimal number. All days in a  |                        |       |
|           | new year preceding the first   |                        |       |
|           | Monday are considered to be in |                        |       |

> |           | week ``'0'``.                  |                        |       |

Reviewer: Ditto:

```suggestion
|           | week 0.                  |                        |       |
```

|           | the ISO week (``%V``).         |                        |       |
+-----------+--------------------------------+------------------------+-------+
| ``%u``    | ISO 8601 weekday as a decimal  | 1, 2, ..., 7           |       |

> |           | number where ``'1'`` is Monday.|                        |       |

Reviewer: I'm not sure about this one.

```suggestion
|           | number where 1 is Monday.|                        |       |
```
